crypto/vrf_go.go

```diff
@@ -0,0 +1,465 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package crypto
+
+import (
+	"crypto/sha512"
+	"crypto/subtle"
+	"fmt"
+
+	"filippo.io/edwards25519"
+	"filippo.io/edwards25519/field"
+)
+
+const (
+	vrfSuite = 0x04 // ECVRF-ED25519-SHA512-Elligator2
+)
+
+// VrfKeygenFromSeedGo deterministically generates a VRF keypair from 32 bytes of (secret) entropy.
+func VrfKeygenFromSeedGo(seed [32]byte) (VrfPubkey, VrfPrivkey) {
+	var pk VrfPubkey
+	var sk VrfPrivkey
+	h := sha512.New()
+	h.Write(seed[:])
+	hSum := h.Sum(nil)
+	copy(sk[:], hSum[:32])
+	sk[0] &= 248
+	sk[31] &= 127
+	sk[31] |= 64
+	p := edwards25519.NewScalar()
+	skBytes := make([]byte, 64)
+	copy(skBytes, sk[:32])
+	p.SetUniformBytes(skBytes)
+	A := edwards25519.NewIdentityPoint().ScalarBaseMult(p)
+	copy(pk[:], A.Bytes())
+	copy(sk[:], seed[:])
+	copy(sk[32:], pk[:])
+	return pk, sk
+}
+
+func (pk VrfPubkey) verifyBytesGo(proof VrfProof, msg []byte) (bool, VrfOutput) {
+	var out VrfOutput
+	h, err := vrfVerifyAndHash(pk[:], proof[:], msg)
+	if err != nil {
+		// TODO: this method should probably return an error.
+		// fmt.Println("issue verifying:", err)
+		return false, out
+	}
+	copy(out[:], h)
+	return true, out
+}
+
+/* Verify a VRF proof (for a given a public key and message) and validate the
+ * public key. If verification succeeds, store the VRF output hash in output[].
+ * Specified in draft spec section 5.3.
+ *
+ * For a given public key and message, there are many possible proofs but only
+ * one possible output hash.
+ */
+func vrfVerifyAndHash(pk []byte, proof []byte, msg []byte) ([]byte, error) {
+	Y := &edwards25519.Point{}
+	// validate key
+	if _, err := Y.SetBytes(pk); err != nil {
+		return nil, err
+	}
+	isSmallOrder := (&edwards25519.Point{}).MultByCofactor(Y).Equal(edwards25519.NewIdentityPoint()) == 1
+	if isSmallOrder {
+		return nil, fmt.Errorf("expected key to have small order")
```

Comment:
Other way around in the error text.

```suggestion
		return nil, fmt.Errorf("public key is a small order point")
```